### PR TITLE
blog: Pixel testing update

### PR DESCRIPTION
**Pull request by mvollmer**
https://mvollmer.github.io/cockpit-project.github.io/blog/pixel-testing-update.html

### CODE REVIEW

**Comment by martinpitt** [comment]
Another blog post is certainly fine. However, I wonder if we shouldn't rather fold this into cockpit's HACKING.md, where we can update it together with the code?

**Comment by mvollmer** [comment]
> I wonder if we shouldn't rather fold this into cockpit's HACKING.md

Which parts exactly? :-)

**Comment by martinpitt** [comment]
Heh, when I wrote that, the entire commit consisted of just the header. This "delta" description is certainly fine.

I was thinking that the operational bits of https://cockpit-project.org/blog/pixel-testing.html (and possible updates since then) should be in HACKING.md -- like "how do I add, update, and locally run a pixel test".

File: _posts/2022-02-03-pixel-tests-update.md [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+Cockpit until we are fully happy with it.  My feeling is that this
+would not have happened without automated pixel testing.
+
+A recent example is [this pull
+request](https://github.com/cockpit-project/cockpit-podman/pull/902).
+We did know about the broken mobile layout of the images list, but the
+thing that kicked us to actually fix it was the enabling of "mobile"
+pixel tests in the cockpit-podman repository.  Having to commit a
+obviously terrible reference image is unpleasant enough to finally sit
+down and fix the damn thing.
```
Comment (martinpitt) [inline _posts/2022-02-03-pixel-tests-update.md:35]: Maybe s/damn// :grin:

File: _posts/2022-02-03-pixel-tests-update.md [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+testing machinery in general.
+
+## Why don't we hate them?
+
+I was expecting that adding pixel tests to Cockpit would be a hard
+sell, but we all seem to like them.  That's awesome!  What a team!
+
+I was afraid that pixel tests would be very unreliable because
+browsers would not be nearly pixel perfect enough.  But it turns out
+that they are very good at this, actually.  We had to switch off font
```
Comment (martinpitt) [inline _posts/2022-02-03-pixel-tests-update.md:105]: Even updating browsers to new versions is surprisingly robust.

File: _posts/2022-02-03-pixel-tests-update.md [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+One thing we have learned is to update reference images only very late
+in the life of a pull request. Resolving merge conflicts of binary
+files in a submodule is just not fun.
+
+I think we all like them because the value that we derive from pixel
+tests is well worth their cost. It feels awesome to officially replace
+an image of a sucky part of Cockpit with a nice one while fixing the
+code, and to show off a new feature with a couple of screenshots that
+you know will be kept from regressing.
```
Comment (martinpitt) [inline _posts/2022-02-03-pixel-tests-update.md:127]: We can now also land PatternFly updates without a human visual inspection and rely on green tests instead, as by now they already do a better judgement than us humans could ever do.

File: _posts/2022-02-03-pixel-tests-update.md [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+So I would say the future looks good for pixel tests.  We are still
+adding more, and we have recently added a whole new browser size with
+it's own set of 77 new reference images.  We keep them working, and
+they keep Cockpit pretty and us honest.
+
+I count that as a success!  And a lot of thanks to the team for being
+so willing to experiment.  Especially to Katerina for being even more
+excited about pixel tests in the beginning than I was. :-)
```
Comment (martinpitt) [inline _posts/2022-02-03-pixel-tests-update.md:136]: Use the official 🙂 emoj?

**Review by martinpitt — COMMENTED** [review]
Nice article! I don't want to be overly pedantic here, as it's a "personal developer" post, but I do have a few suggestions.

File: _posts/2022-02-03-pixel-tests-update.md (hunk 1 above)
Comment (mvollmer) [reply _posts/2022-02-03-pixel-tests-update.md:35]: Yep, no objections.

File: _posts/2022-02-03-pixel-tests-update.md (hunk 4 above)
Comment (mvollmer) [reply _posts/2022-02-03-pixel-tests-update.md]: Done.

File: _posts/2022-02-03-pixel-tests-update.md (hunk 3 above)
Comment (mvollmer) [reply _posts/2022-02-03-pixel-tests-update.md:127]: Added.

File: _posts/2022-02-03-pixel-tests-update.md (hunk 2 above)
Comment (mvollmer) [reply _posts/2022-02-03-pixel-tests-update.md]: Added.

**Review by martinpitt — APPROVED** [review]
Nice! Send it to the ether, and I'll tweet it from the cockpit account.

**Pull request merged**